packages/core/echo/echo-db/src/hypergraph.ts

```diff
@@ -235,25 +239,104 @@ export type GraphQueryContextParams = {
   onStop: () => void;
 };
 
+/**
+ * Query data source.
+ * Implemented by a space or a remote agent.
+ * Each query has a separate instance.
+ */
+export interface QuerySource {
+  // TODO(dmaretskyi): Update info?
+  changed: Event<void>;
+
+  getResults(): QueryResult[];
+
+  /**
+   * One-shot query.
+   */
+  run(filter: Filter): Promise<QueryResult[]>;
+
+  /**
+   * Set the filter and trigger continuous updates.
+   */
+  update(filter: Filter): void;
+
+  // TODO(dmaretskyi): Make async.
+  close(): void;
```

Comment:
I think the behavior is not clear without a matching start() method.
I'd rename update to subsribe and close to clearSubscriptions or something like that